File: src/main/java/komu/blunt/parser/Parser.java
package komu.blunt.parser;

import com.google.common.collect.ImmutableList;
import komu.blunt.ast.*;
import komu.blunt.objects.Symbol;
import komu.blunt.types.DataTypeDefinitions;
import komu.blunt.types.patterns.Pattern;
import komu.blunt.types.patterns.VariablePattern;

import java.util.ArrayList;
import java.util.List;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static komu.blunt.objects.Symbol.symbol;
import static komu.blunt.parser.Associativity.LEFT;
import static komu.blunt.parser.TokenType.*;

public final class Parser {

    private final Lexer lexer;
    private final OperatorSet operators = new OperatorSet();
    private final TypeParser typeParser;
    private final PatternParser patternParser;
    private final DataTypeParser dataTypeParser;

    @SuppressWarnings("unchecked")
    private static final List<TokenType<?>> expressionStartTokens =
        asList(IF, LET, LAMBDA, LPAREN, LBRACKET, LITERAL, IDENTIFIER, TYPE_OR_CTOR_NAME, CASE);

    public Parser(String source) {
        this.lexer = new Lexer(source, operators);
        this.typeParser = new TypeParser(lexer);
        this.patternParser = new PatternParser(lexer);
        this.dataTypeParser = new DataTypeParser(lexer, typeParser);
    }

    public static ASTExpression parseExpression(String source) {
        return new Parser(source).parseExpression();
    }
    
    public List<ASTDefinition> parseDefinitions() {
        List<ASTDefinition> result = new ArrayList<>();
        
        while (lexer.hasMoreTokens())
            result.add(parseDefinition());
        
        return result;
    }
    
    private ASTDefinition parseDefinition() {
        if (lexer.nextTokenIs(DATA))
            return dataTypeParser.parseDataDefinition();
        else
            return parseValueDefinition();
    }


    private ASTValueDefinition parseValueDefinition() {
        LexerState lexerState = lexer.save();
        try {
            try {
                return parseSimpleDefinition();
            } catch (SyntaxException e) {
                lexer.restore(lexerState);
                return parseOperatorDefinition();
            }

        } catch (SyntaxException e) {
            lexer.restore(lexerState);
            return parseNormalValueDefinition();
        }
    }

    private ASTValueDefinition parseSimpleDefinition() {
        lexer.pushBlockStartAtNextToken();

        Symbol name = parseIdentifier();

        lexer.expectToken(ASSIGN);
        ASTExpression value = parseExpression();
        lexer.expectToken(END);
        return AST.define(name, value);
    }

    // <pattern> <op> <pattern> = <exp> ;;
    private ASTValueDefinition parseOperatorDefinition() {
        lexer.pushBlockStartAtNextToken();
        
        Pattern left = patternParser.parseSimplePattern();
        Operator op = lexer.readTokenValue(OPERATOR);
        Pattern right = patternParser.parseSimplePattern();
     
        lexer.expectToken(ASSIGN);

        ASTExpression value = parseExpression();
        lexer.expectToken(END);

        FunctionBuilder functionBuilder = new FunctionBuilder();
        functionBuilder.addAlternative(ImmutableList.of(left, right), value);
        return AST.define(op.toSymbol(), functionBuilder.build());
    }
    
    // <ident> <pattern>+ = <exp> ;;
    private ASTValueDefinition parseNormalValueDefinition() {
        FunctionBuilder functionBuilder = new FunctionBuilder();

        Symbol name = null;
        while (name == null || nextTokenIsIdentifier(name)) {
            lexer.pushBlockStartAtNextToken();
            name = parseIdentifier();

            ImmutableList.Builder<Pattern> argsBuilder = ImmutableList.builder();

            while (!lexer.nextTokenIs(ASSIGN))
                argsBuilder.add(patternParser.parseSimplePattern());

            lexer.expectToken(ASSIGN);

            ASTExpression value = parseExpression();
            lexer.expectToken(END);
            functionBuilder.addAlternative(argsBuilder.build(), value);
        }

        return AST.define(name, functionBuilder.build());
    }

    private boolean nextTokenIsIdentifier(Symbol name) {
        return lexer.nextTokenIs(IDENTIFIER) && lexer.peekTokenValue(IDENTIFIER).equals(name.toString());
    }

    public ASTExpression parseExpression() {
        ASTExpression exp = parseExp(0);

        while (true) {
            if (lexer.readMatchingToken(SEMICOLON)) {
                ASTExpression rhs = parseExp(0);
                exp = AST.sequence(exp, rhs);
            } else {
                return exp;
            }
        }
    }

    private ASTExpression parseExp(int level) {
        if (level <= operators.getMaxLevel())
            return parseExpN(level);
        else
            return parseApplicative();
    }

    private ASTExpression parseExpN(int level) {
        ASTExpression exp = parseExp(level + 1);

        while (true) {
            Operator op = lexer.readOperatorMatchingLevel(level);
            if (op != null) {
                ASTExpression rhs = parseExp(op.associativity == LEFT ? level+1 : level);
                exp = binary(op, exp, rhs);
            } else {
                return exp;
            }
        }
    }

    private ASTExpression parseApplicative() {
        ASTExpression exp = parsePrimitive();

        while (lexer.nextTokenIsOneOf(expressionStartTokens))
            exp = AST.apply(exp, parsePrimitive());

        return exp;
    }

    private ASTExpression parsePrimitive() {
        TokenType type = lexer.peekTokenType();
        
        if (type == EOF)
            throw parseError("unexpected eof");
        
        if (type == IF)
            return parseIf();
        
        if (type == LET)
            return parseLet();
            
        if (type == LAMBDA)
            return parseLambda();
        
        if (type == CASE)
            return parseCase();
        
        if (type == LPAREN)
            return parseParens();
        
        if (type == LBRACKET)
            return parseList();

        if (type == LITERAL)
            return AST.constant(lexer.readTokenValue(LITERAL));

        return parseVariableOrConstructor();
    }

    // if <expr> then <expr> else <expr>
    private ASTExpression parseIf() {
        lexer.expectToken(IF);
        ASTExpression test = parseExpression();
        lexer.expectToken(THEN);
        ASTExpression cons = parseExpression();
        lexer.expectToken(ELSE);
        ASTExpression alt = parseExpression();

        return AST.ifExp(test, cons, alt);
    }

    // case <exp> of <alternative>+
    private ASTExpression parseCase() {
        lexer.expectIndentStartToken(CASE);
        ASTExpression exp = parseExpression();
        lexer.expectToken(OF);

        ImmutableList.Builder<ASTAlternative> alts = ImmutableList.builder();
        do {
            alts.add(parseAlternative());
        } while (!lexer.nextTokenIs(END));
        lexer.expectToken(END);

        return AST.caseExp(exp, alts.build());
    }

    private ASTAlternative parseAlternative() {
        Pattern pattern = patternParser.parsePattern();
        lexer.expectIndentStartToken(RIGHT_ARROW);
        ASTExpression exp = parseExpression();
        lexer.expectToken(END);
        return AST.alternative(pattern, exp);
    }

    // let [rec] <ident> <ident>* = <expr> in <expr>
    private ASTExpression parseLet() {
        lexer.expectToken(LET);
        boolean recursive = lexer.readMatchingToken(REC);
        
        Symbol name = parseIdentifier();

        List<Symbol> args = new ArrayList<>();

        if (lexer.nextTokenIs(OPERATOR)) {
            Operator op = lexer.readTokenValue(OPERATOR);
            args.add(name);
            args.add(parseIdentifier());
            name = symbol(op.toString());

            lexer.expectToken(ASSIGN);
        } else {
            while (!lexer.readMatchingToken(ASSIGN))
                args.add(parseIdentifier());
        }

        ASTExpression value = parseExpression();
        lexer.expectToken(IN);
        ASTExpression body = parseExpression();

        if (!args.isEmpty())
            value = AST.lambda(args, value);

        return AST.let(recursive, new ImplicitBinding(name, value), body);
    }

    // \ <ident> -> expr
    private ASTExpression parseLambda() {
        lexer.expectToken(LAMBDA);

        ImmutableList.Builder<Pattern> argsBuilder = ImmutableList.builder();

        do {
            argsBuilder.add(patternParser.parseSimplePattern());
        } while (!lexer.readMatchingToken(RIGHT_ARROW));
        
        FunctionBuilder builder = new FunctionBuilder();
        ImmutableList<Pattern> patterns = argsBuilder.build();
        builder.addAlternative(patterns, parseExpression());
        return builder.build();
    }

    // () | (<op>) | ( <expr> )
    private ASTExpression parseParens() {
        lexer.expectToken(LPAREN);
        if (lexer.readMatchingToken(RPAREN))
            return AST.constructor(DataTypeDefinitions.UNIT);
        
        if (lexer.nextTokenIs(OPERATOR)) {
            Operator op = lexer.readTokenValue(OPERATOR);
            lexer.expectToken(RPAREN);

            return op.isConstructor() ? AST.constructor(op.toString()) : AST.variable(op.toString());
        }

        List<ASTExpression> exps = new ArrayList<>();
        
        do {
            exps.add(parseExpression());    
        } while (lexer.readMatchingToken(COMMA));

        lexer.expectToken(RPAREN);

        if (exps.size() == 1)
            return exps.get(0);
        else
            return AST.tuple(exps);
    }
    
    // []
    // [<exp> (,<exp>)*]
    private ASTExpression parseList() {
        AST.ListBuilder list = AST.listBuilder();

        lexer.expectToken(LBRACKET);

        if (!lexer.nextTokenIs(RBRACKET)) {
            do {
                list.add(parseExpression());
            } while (lexer.readMatchingToken(COMMA));
        }

        lexer.expectToken(RBRACKET);

        return list.build();
    }

    private ASTExpression parseVariableOrConstructor() {
        if (lexer.nextTokenIs(IDENTIFIER))
            return AST.variable(lexer.readTokenValue(IDENTIFIER));

        if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME))
            return AST.constructor(lexer.readTokenValue(TYPE_OR_CTOR_NAME));

        if (lexer.readMatchingToken(LPAREN)) {
            Operator op = lexer.readTokenValue(OPERATOR);
            lexer.expectToken(RPAREN);

            if (op.isConstructor())
                return AST.constructor(op.toString());
            else
                return AST.variable(op.toString());
        }

        throw lexer.expectFailure("identifier or type constructor");
    }

    private Symbol parseIdentifier() {
        if (lexer.nextTokenIs(IDENTIFIER)) {
            return symbol(lexer.readTokenValue(IDENTIFIER));
        }

        if (lexer.readMatchingToken(LPAREN)) {
            Operator op = lexer.readTokenValue(OPERATOR);
            lexer.expectToken(RPAREN);
            return op.toSymbol();
        }

        throw lexer.expectFailure("identifier");
    }

    private SyntaxException parseError(final String s) {
        return lexer.parseError(s);
    }

    private static ASTExpression binary(Operator op, ASTExpression lhs, ASTExpression rhs) {
        ASTExpression exp = op.isConstructor() ? AST.constructor(op.toString()) : AST.variable(op.toString());
        
        return AST.apply(exp, lhs, rhs);
    }
    
    private static class FunctionBuilder {
        private final List<Symbol> symbols = new ArrayList<>();
        private final List<ASTExpression> exps = new ArrayList<>();
        private final ImmutableList.Builder<ASTAlternative> alternatives = ImmutableList.builder();
        
        void addAlternative(ImmutableList<Pattern> args, ASTExpression body) {
            if (exps.isEmpty()) {
                for (int i = 0; i < args.size(); i++) {
                    Symbol var = symbol("$arg" + i); // TODO: fresh symbols
                    symbols.add(var);
                    exps.add(AST.variable(var));
                }                            
            } else if (args.size() != exps.size()) {
                throw new SyntaxException("invalid amount of arguments");
            }
            
            alternatives.add(AST.alternative(Pattern.tuple(args), body));
        }
        
        ASTExpression build() {
            ImmutableList<ASTAlternative> alts = alternatives.build();
            
            // optimization
            List<Symbol> simpleVars = containsOnlyVariablePatterns(alts);
            if (simpleVars != null)
                return AST.lambda(simpleVars, alts.get(0).value);

            return AST.lambda(symbols, AST.caseExp(AST.tuple(exps), alts));
        }

        private List<Symbol> containsOnlyVariablePatterns(ImmutableList<ASTAlternative> alts) {
            if (alts.size() == 1)
                return variablePattern(alts.get(0).pattern);
            return null;
        }

        private List<Symbol> variablePattern(Pattern pattern) {
            if (pattern instanceof VariablePattern)
                return singletonList(((VariablePattern) pattern).var);

            /*
            if (pattern instanceof ConstructorPattern) {
                ConstructorPattern c = (ConstructorPattern) pattern;

                if (c.name.equals("()"))
                    return null;
                if (!c.name.equals(DataTypeDefinitions.tupleName(c.args.size())))
                    return null;

                List<Symbol> vars = new ArrayList<>(c.args.size());

                for (Pattern p : c.args)
                    if (p instanceof VariablePattern) {
                        vars.add(((VariablePattern) p).var);
                    } else {
                        return null;
                    }

                return vars;
            }
            */

            return null;
        }
    }
}

File: src/main/kotlin/komu/blunt/parser/Parser.kt
package komu.blunt.parser

import komu.blunt.ast.*
import komu.blunt.objects.Symbol
import komu.blunt.parser.TokenType.Companion.ASSIGN
import komu.blunt.parser.TokenType.Companion.CASE
import komu.blunt.parser.TokenType.Companion.DATA
import komu.blunt.parser.TokenType.Companion.ELSE
import komu.blunt.parser.TokenType.Companion.END
import komu.blunt.parser.TokenType.Companion.EOF
import komu.blunt.parser.TokenType.Companion.IDENTIFIER
import komu.blunt.parser.TokenType.Companion.IF
import komu.blunt.parser.TokenType.Companion.LAMBDA
import komu.blunt.parser.TokenType.Companion.LBRACKET
import komu.blunt.parser.TokenType.Companion.LET
import komu.blunt.parser.TokenType.Companion.LITERAL
import komu.blunt.parser.TokenType.Companion.LPAREN
import komu.blunt.parser.TokenType.Companion.OF
import komu.blunt.parser.TokenType.Companion.OPERATOR
import komu.blunt.parser.TokenType.Companion.RBRACKET
import komu.blunt.parser.TokenType.Companion.REC
import komu.blunt.parser.TokenType.Companion.RIGHT_ARROW
import komu.blunt.parser.TokenType.Companion.RPAREN
import komu.blunt.parser.TokenType.Companion.SEMICOLON
import komu.blunt.parser.TokenType.Companion.THEN
import komu.blunt.parser.TokenType.Companion.TYPE_OR_CTOR_NAME
import komu.blunt.types.ConstructorNames

fun parseExpression(source: String) =
    Parser(source).parseExpression()

class Parser(source: String) {

    private val operators = OperatorSet()
    private val lexer = Lexer(source, operators)
    private val typeParser = TypeParser(lexer)
    private val patternParser = PatternParser(lexer)
    private val dataTypeParser = DataTypeParser(lexer, typeParser)

    private val EXPRESSION_START_TOKENS =
        listOf(IF, LET, LAMBDA, LPAREN, LBRACKET, LITERAL, IDENTIFIER, TYPE_OR_CTOR_NAME, CASE)

    fun parseDefinitions(): List<ASTDefinition> =
        lexer.listOf {
            if (lexer.hasMoreTokens())
                parseDefinition()
            else
                null
        }

    fun parseDefinition(): ASTDefinition =
        if (lexer.nextTokenIs(DATA))
            dataTypeParser.parseDataDefinition()
        else
            parseValueDefinition()

    private fun parseValueDefinition(): ASTValueDefinition = when {
        isSimpleDefinition() ->
            parseSimpleDefinition()
        isOperatorDefinition() ->
            parseOperatorDefinition()
        else ->
            parseNormalValueDefinition()
    }

    private fun isSimpleDefinition(): Boolean =
        lexer.withSavedState {
            try {
                parseIdentifier()
                lexer.nextTokenIs(ASSIGN)
            } catch (e: SyntaxException) {
                false
            }
        }

    private fun isOperatorDefinition(): Boolean =
        lexer.withSavedState {
            try {
                patternParser.parseSimplePattern()
                lexer.nextTokenIs(OPERATOR)
            } catch (e: SyntaxException) {
                false
            }
        }

    private fun parseSimpleDefinition(): ASTValueDefinition {
        lexer.pushBlockStartAtNextToken()

        val name = parseIdentifier()

        lexer.expectToken(ASSIGN)
        val value = parseExpression()
        lexer.expectToken(END)
        return ASTValueDefinition(name, value)
    }

    // <pattern> <op> <pattern> = <exp> ;;
    private fun parseOperatorDefinition(): ASTValueDefinition {
        lexer.pushBlockStartAtNextToken()

        val left = patternParser.parseSimplePattern()
        val op = lexer.readTokenValue(OPERATOR)
        val right = patternParser.parseSimplePattern()

        lexer.expectToken(ASSIGN)

        val value = parseExpression()
        lexer.expectToken(END)

        return ASTValueDefinition(op.toSymbol(), buildFunction(listOf(left, right), value))
    }

    // <ident> <pattern>+ = <exp> ;;
    private fun parseNormalValueDefinition(): ASTValueDefinition {
        val functionBuilder = FunctionBuilder()

        var name: Symbol? = null
        while (name == null || nextTokenIsIdentifier(name)) {
            lexer.pushBlockStartAtNextToken()
            name = parseIdentifier()

            val args = lexer.listOf1 {
                if (lexer.nextTokenIs(ASSIGN))
                    null
                else
                    patternParser.parseSimplePattern()
            }

            lexer.expectToken(ASSIGN)

            val value = parseExpression()
            lexer.expectToken(END)
            functionBuilder.addAlternative(args, value)
        }

        return ASTValueDefinition(name, functionBuilder.build())
    }

    private fun nextTokenIsIdentifier(name: Symbol) =
        lexer.nextTokenIs(IDENTIFIER) && lexer.peekTokenValue(IDENTIFIER) == name.toString()

    fun parseExpression(): ASTExpression {
        var exp = parseExp(0)

        while (true) {
            if (lexer.readMatchingToken(SEMICOLON)) {
                val rhs = parseExp(0)
                exp = ASTExpression.Sequence(exp, rhs)
            } else {
                return exp
            }
        }
    }

    private fun parseExp(level: Int): ASTExpression =
        if (level <= operators.maxLevel)
            parseExpN(level)
        else
            parseApplicative()

    private fun parseExpN(level: Int): ASTExpression {
        var exp = parseExp(level + 1)

        while (true) {
            val op = lexer.readOperatorMatchingLevel(level) ?: return exp

            val rhs = parseExp(if (op.associativity == Associativity.LEFT) level + 1 else level)
            exp = binary(op, exp, rhs)
        }
    }

    private fun parseApplicative(): ASTExpression {
        var exp = parsePrimitive()

        while (lexer.nextTokenIsOneOf(EXPRESSION_START_TOKENS))
            exp = AST.apply(exp, parsePrimitive())

        return exp
    }

    private fun parsePrimitive(): ASTExpression =
        when (lexer.peekTokenType()) {
            EOF         -> parseError("unexpected eof")
            IF          -> parseIf()
            LET         -> parseLet()
            LAMBDA      -> parseLambda()
            CASE        -> parseCase()
            LPAREN      -> parseParens()
            LBRACKET    -> parseList()
            LITERAL     -> ASTExpression.Constant(lexer.readTokenValue(LITERAL))
            else        -> parseVariableOrConstructor()
        }

    // if <expr> then <expr> else <expr>
    private fun parseIf(): ASTExpression {
        lexer.expectToken(IF)
        val test = parseExpression()
        lexer.expectToken(THEN)
        val cons = parseExpression()
        lexer.expectToken(ELSE)
        val alt = parseExpression()

        return AST.ifExp(test, cons, alt)
    }

    // case <exp> of <alternative>+
    private fun parseCase(): ASTExpression {
        lexer.expectIndentStartToken(CASE)
        val exp = parseExpression()
        lexer.expectToken(OF)

        val alts = lexer.listOf1 {
            if (lexer.nextTokenIs(END))
                null
            else
                parseAlternative()
        }
        lexer.expectToken(END)

        return ASTExpression.Case(exp, alts)
    }

    private fun parseAlternative(): ASTAlternative {
        val pattern = patternParser.parsePattern()
        lexer.expectIndentStartToken(RIGHT_ARROW)
        val exp = parseExpression()
        lexer.expectToken(END)
        return ASTAlternative(pattern, exp)
    }

    // let [rec] <ident> <ident>* = <expr> in <expr>
    private fun parseLet(): ASTExpression {
        lexer.expectToken(LET)
        val recursive = lexer.readMatchingToken(REC)

        var name = parseIdentifier()

        val args = if (lexer.nextTokenIs(OPERATOR)) {
            name = lexer.readTokenValue(OPERATOR).toSymbol()
            val ident = parseIdentifier()

            lexer.expectToken(ASSIGN)
            listOf(name, ident)
        } else {
            lexer.listOf {
                if (lexer.readMatchingToken(ASSIGN))
                    null
                else
                    parseIdentifier()
            }
        }

        var value = parseExpression()
        lexer.expectToken(TokenType.IN)
        val body = parseExpression()

        if (args.any())
            value = AST.lambda(args, value)

        return AST.let(recursive, ImplicitBinding(name, value), body)
    }

    // \ <ident> -> expr
    private fun parseLambda(): ASTExpression {
        lexer.expectToken(LAMBDA)

        val args = lexer.listOf1 {
            if (lexer.readMatchingToken(RIGHT_ARROW))
                null
            else
                patternParser.parseSimplePattern()
        }

        return buildFunction(args, parseExpression())
    }

    // () | (<op>) | ( <expr> )
    private fun parseParens(): ASTExpression =
        lexer.inParens {
            when {
                lexer.nextTokenIs(RPAREN) ->
                    AST.constructor(ConstructorNames.UNIT)

                lexer.nextTokenIs(OPERATOR) ->
                    operatorExp(lexer.readTokenValue(OPERATOR))

                else -> {
                    val exps = lexer.commaSep { parseExpression() }
                    exps.singleOrNull() ?: ASTExpression.tuple(exps)
                }
            }
        }

    // []
    // [<exp> (,<exp>)*]
    private fun parseList(): ASTExpression =
        lexer.inBrackets {
            if (!lexer.nextTokenIs(RBRACKET))
                ASTExpression.list(lexer.commaSep { parseExpression() })
            else
                ASTExpression.list(emptyList())
        }

    private fun parseVariableOrConstructor(): ASTExpression = when {
        lexer.nextTokenIs(IDENTIFIER) ->
            ASTExpression.Variable(lexer.readTokenValue(IDENTIFIER))

        lexer.nextTokenIs(TYPE_OR_CTOR_NAME) ->
            AST.constructor(lexer.readTokenValue(TYPE_OR_CTOR_NAME))

        lexer.nextTokenIs(LPAREN) ->
            lexer.inParens { operatorExp(lexer.readTokenValue(OPERATOR)) }

        else -> lexer.expectFailure("identifier or type constructor")
    }

    private fun parseIdentifier(): Symbol = when {
        lexer.nextTokenIs(IDENTIFIER) ->
            Symbol(lexer.readTokenValue(IDENTIFIER))
        lexer.nextTokenIs(LPAREN) ->
            lexer.inParens { lexer.readTokenValue(OPERATOR).toSymbol() }
        else ->
            lexer.expectFailure("identifier")
    }

    private fun parseError(s: String): Nothing =
        lexer.parseError(s)

    private fun binary(op: Operator, lhs: ASTExpression, rhs: ASTExpression) =
        AST.apply(operatorExp(op), lhs, rhs)

    private fun operatorExp(op: Operator): ASTExpression =
        if (op.isConstructor) AST.constructor(op.toString()) else ASTExpression.Variable(op.toSymbol())
}
